File: apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/RecordSet.java
package com.apollographql.android.cache.normalized;

import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

public final class RecordSet {

  private final Map<String, Record> recordMap = new LinkedHashMap<>();

  public Record get(String key) {
    return recordMap.get(key);
  }

  public Set<String> merge(Record apolloRecord) {
    final Record oldRecord = recordMap.get(apolloRecord.key());
    if (oldRecord == null) {
      recordMap.put(apolloRecord.key(), apolloRecord);
      return Collections.emptySet();
    } else {
      return oldRecord.mergeWith(apolloRecord);
    }
  }

  public Collection<Record> allRecords() {
    return recordMap.values();
  }
}

File: apollo-normalized-cache-api/src/main/java/com/apollographql/apollo/cache/normalized/RecordSet.kt
package com.apollographql.apollo.cache.normalized

class RecordSet {
  private val recordMap = mutableMapOf<String, Record>()

  operator fun get(key: String): Record? = recordMap[key]

  fun merge(apolloRecord: Record): Set<String> {
    val oldRecord = recordMap[apolloRecord.key]
    return if (oldRecord == null) {
      recordMap[apolloRecord.key] = apolloRecord
      emptySet()
    } else {
      oldRecord.mergeWith(apolloRecord)
    }
  }

  fun allRecords(): Collection<Record> = recordMap.values.toList()
}
